feat: add create-spacetime CLI tool

clockwork-tien opened the pull request:
# Description of Changes
- Introduce CLI tool for scaffolding SpacetimeDB projects and create applications with React frontends and Rust or C# backends
- Add `templates/create-spacetime/` with complete CLI implementation which supports interactive project creation, package manager detection, and deployment options (Maincloud and local deployment)
- Update `pnpm-workspace.yaml` to include new package
- Detailed package description and usage examples provided in `templates/create-spacetime/README.md`

# Screenshots
- Run CLI tool using `npm create spacetime@latest`
<img width="665" height="271" alt="image" src="https://github.com/user-attachments/assets/410f19f5-4143-46e4-88ef-9e2443196899" />
<img width="460" height="220" alt="image" src="https://github.com/user-attachments/assets/3ee59845-5d0b-4fe5-85c1-b9fae1f13e05" />

# Testing
- [x] Package builds successfully (`pnpm run build` within `templates/create-spacetime`) and runs without errors (`pnpm run test`)
- [x] CLI creates projects with interactive prompts, which support both Rust and C# templates with React and deployment options for local or Maincloud

clockwork-tien committed:
feat: create-spacetime command and package

clockwork-tien committed:
feat(create-spacetime): general improvements and refactor

clockwork-tien committed:
feat(create-spacetime): deployment config improvements, code refactor

clockwork-tien committed:
Merge branch 'master' of https://github.com/clockworklabs/SpacetimeDB into tien/feat/create-spacetime

cloutiertyler commented on templates/create-spacetime/.eslintrc.json:
Please update this to the newer `eslint.config.js` format.

cloutiertyler replied:
If this eslint config is just for us, we could also just rely on the root eslint.config.js instead as well.

cloutiertyler commented on templates/create-spacetime/src/index.ts:
```diff
@@ -0,0 +1,291 @@
+    console.log();
+  }
+
+  if (options.template && !isValidTemplate(options.template)) {
```
Would it be possible to modify the script so that if it's not in the list of built in templates, it tries to clone the template from GitHub?

clockwork-tien replied:
Thank you, I have updated to add support/option for non built-in templates.

cloutiertyler reviewed (commented):
This is looking really really good. I just left one more request about downloading templates from GitHub if the user specifies one that we don't know about.

Also as a minor nit, `create-spacetime` itself is not really a template, so it probably shouldn't go in the templates directory. I would imagine the templates directory is where we store/host all the built in templates.

clockwork-tien committed:
feat(create-spacetime): custom template option, move to cli folder, use eslint.config.js format

clockwork-tien committed:
Merge branch 'master' of https://github.com/clockworklabs/SpacetimeDB into tien/feat/create-spacetime

clockwork-tien committed:
feat(create-spacetime): update to use root eslint.config.js

clockwork-tien commented:
@cloutiertyler Thank you for the review! I have updated to use root `eslint.config.js`, implemented support/option for non built-in templates, and moved `create-spacetime` from `templates` folder to `cli` folder.

cloutiertyler commented on cli/create-spacetime/src/constants.ts:
```diff
@@ -0,0 +1,20 @@
+  SERVER_START_DELAY: 3000,
+} as const;
+
+export const SPACETIME_SDK_PACKAGE = "@clockworklabs/spacetimedb-sdk";
```
This has now been updated to just `spacetimedb`
```suggestion
export const SPACETIME_SDK_PACKAGE = "spacetimedb";
```

https://www.npmjs.com/package/spacetimedb

cloutiertyler commented on cli/create-spacetime/src/constants.ts:
```diff
@@ -0,0 +1,20 @@
+
+export const SPACETIME_VERSIONS = {
+  SDK: "^1.3.1",
+  RUNTIME: "1.3.*",
```
This doesn't appear to be used anywhere.

clockwork-tien replied:
Thank you, I removed this unused variable.

cloutiertyler commented on cli/create-spacetime/src/lib/create-project.ts:
```diff
@@ -0,0 +1,476 @@
+      );
+      content = content.replace(
+        /spacetimedb-lib = \{ path = ".*" \}/g,
+        `spacetimedb-lib = "${SPACETIME_VERSIONS.CLI}"`,
```
This is not the CLI version, it should probably be the runtime version?

cloutiertyler commented on cli/create-spacetime/src/index.ts:
```diff
@@ -0,0 +1,316 @@
+    process.exit(1);
+  }
+
+  if (!deps.spacetime) {
```
Is there a reason we couldn't just install this for them?

cloutiertyler replied:
Eventually I think we'd like to move most of this functionality directly into the `spacetime` cli command, so it would make sense that this actually just becomes a simple bootstrapping TypeScript script to install the CLI and then run a command like:

`spacetime init --template foobar/baz`

cloutiertyler commented on cli/create-spacetime/tsconfig.json:
Could you rename `cli` to `npm-packages` instead? Sorry for the back and forth on this.

cloutiertyler commented on cli/create-spacetime/src/lib/create-project.ts:
```diff
@@ -0,0 +1,476 @@
+    clientPackage.name = `${name}-client`;
+
+    if (clientPackage.dependencies?.[SPACETIME_SDK_PACKAGE]) {
+      clientPackage.dependencies[SPACETIME_SDK_PACKAGE] = SPACETIME_VERSIONS.SDK;
```
I'm also not sure why we're editing the template to add our own version in here as well. I guess to get them on the most up to date version?

clockwork-tien replied:
The template has `"spacetimedb": "workspace:*"` in package.json, which didn't work when cloning standalone version so I needed to update with a fixed version.

cloutiertyler reviewed (commented):
I'm having the terrible realization that most of this code should really exist within [`spacetime init`](https://github.com/clockworklabs/SpacetimeDB/blob/master/crates/cli/src/subcommands/init.rs) and be written in Rust. That way people who don't want to execute npm or npx can still use the template creator.

This tool would then become a thin wrapper which would essentially just install `spacetime` if it was not there and then execute `spacetime init --template`.

clockwork-tien committed:
Update cli/create-spacetime/src/constants.ts

Co-authored-by: Tyler Cloutier <cloutiertyler@users.noreply.github.com>
Signed-off-by: clockwork-tien <tien@clockworklabs.io>

clockwork-tien committed:
feat(create-spacetime): move to npm-packages folder, updates to use spacetimedb package

clockwork-tien committed:
chore(create-spacetime): update SPACETIME_VERSIONS var namings
